Bug: MemberExpression.object should be Expression, not LeftHandSideExpression

### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

typescript-estree

### Playground Link

_No response_

### Repro Code

```typescript
function isExpressionShould(node: TSESTree.MemberExpression): boolean {
    const object = node.object;
    const property = node.property;

    return (
        object?.type === AST_NODE_TYPES.BinaryExpression &&
        isIdentifier(property) &&
        property.name === "should"
    );
}
```


### ESLint Config

```javascript
N/A
```


### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```


### Expected Result

I expect my code to compile without errors

### Actual Result

 error TS2367: This condition will always return 'false' since the types 'AST_NODE_TYPES.ArrayExpression | AST_NODE_TYPES.ArrayPattern | AST_NODE_TYPES.ArrowFunctionExpression | AST_NODE_TYPES.CallExpression | AST_NODE_TYPES.ClassExpression | AST_NODE_TYPES.FunctionExpression | ... 15 more ... | AST_NODE_TYPES.TSTypeAssertion' and 'AST_NODE_TYPES.BinaryExpression' have no overlap.

57         object?.type === AST_NODE_TYPES.BinaryExpression &&
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Found 1 error.

### Additional Info

I am working on a repo which uses node 14.16.0 and cannto upgrade to node 14.17.0 (yet) because it is used in an electron 12 project.

I'm trying to write a lint rule to catch sillinesses like `should(a === b).be.true()` etc, and another way it can be written is `(a === b).should.be.true()`, however I cannot get this check to compile which is a problem (I'll probably have to patch the types in my repo to get it working for now)

I checked this code with ast explorer

```ts
(apple === banana).should.be.true();
```

and I get this
![image](https://user-images.githubusercontent.com/5262051/206760188-f310cae5-edeb-493a-8182-d0b121e22acd.png)

I've also caught it in a vscode debugger with the values

<img width="880" alt="image" src="https://user-images.githubusercontent.com/5262051/206760990-cbfe31e3-4495-457d-8077-fff938c02ced.png">

<img width="464" alt="image" src="https://user-images.githubusercontent.com/5262051/206761270-bb100477-b733-4043-b851-e689e186aba7.png">




### Versions

<!-- Fill in the relevant versions and delete the others -->
| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `4.33.0` |
| `@typescript-eslint/parser`            | `4.31.2` |
| `@typescript-eslint/scope-manager`     | `4.33.0` |
| `@typescript-eslint/typescript-estree` | 4.31.2` |
| `@typescript-eslint/type-utils`        | N/A ??? |
| `@typescript-eslint/visitor-keys`        | `4.31.2` |
| `@typescript-eslint/types`             | `4.31.2` |
| `TypeScript`                           | `4.2.3` |
| `ESLint`                               | `7.26.0` |
| `node`                                 | `14.16.0` |

--- a/packages/ast-spec/src/expression/ArrayExpression/spec.ts
+++ b/packages/ast-spec/src/expression/ArrayExpression/spec.ts
@@ -5,5 +5,8 @@ import type { Expression } from '../../unions/Expression';
 
 export interface ArrayExpression extends BaseNode {
   type: AST_NODE_TYPES.ArrayExpression;
-  elements: (Expression | SpreadElement)[];
+  /**
+   * an element will be `null` in the case of a sparse array: `[1, ,3]`
+   */
+  elements: (Expression | SpreadElement | null)[];
 }

--- a/packages/ast-spec/src/expression/MemberExpression/spec.ts
+++ b/packages/ast-spec/src/expression/MemberExpression/spec.ts
@@ -2,11 +2,10 @@ import type { AST_NODE_TYPES } from '../../ast-node-types';
 import type { BaseNode } from '../../base/BaseNode';
 import type { PrivateIdentifier } from '../../special/PrivateIdentifier/spec';
 import type { Expression } from '../../unions/Expression';
-import type { LeftHandSideExpression } from '../../unions/LeftHandSideExpression';
 import type { Identifier } from '../Identifier/spec';
 
 interface MemberExpressionBase extends BaseNode {
-  object: LeftHandSideExpression;
+  object: Expression;
   property: Expression | Identifier | PrivateIdentifier;
   computed: boolean;
   optional: boolean;

--- a/packages/ast-spec/src/unions/ObjectLiteralElement.ts
+++ b/packages/ast-spec/src/unions/ObjectLiteralElement.ts
@@ -1,8 +1,7 @@
-import type { MethodDefinition } from '../element/MethodDefinition/spec';
 import type { Property } from '../element/Property/spec';
 import type { SpreadElement } from '../element/SpreadElement/spec';
 
-export type ObjectLiteralElement = MethodDefinition | Property | SpreadElement;
+export type ObjectLiteralElement = Property | SpreadElement;
 
 // TODO - breaking change remove this
 export type ObjectLiteralElementLike = ObjectLiteralElement;

--- a/packages/eslint-plugin-internal/src/rules/plugin-test-formatting.ts
+++ b/packages/eslint-plugin-internal/src/rules/plugin-test-formatting.ts
@@ -197,8 +197,11 @@ export default createRule<Options, MessageIds>({
       }
     }
 
-    function checkExpression(node: TSESTree.Node, isErrorTest: boolean): void {
-      switch (node.type) {
+    function checkExpression(
+      node: TSESTree.Node | null,
+      isErrorTest: boolean,
+    ): void {
+      switch (node?.type) {
         case AST_NODE_TYPES.Literal:
           checkLiteral(node, isErrorTest);
           break;
@@ -478,7 +481,7 @@ export default createRule<Options, MessageIds>({
 
     function checkValidTest(tests: TSESTree.ArrayExpression): void {
       for (const test of tests.elements) {
-        switch (test.type) {
+        switch (test?.type) {
           case AST_NODE_TYPES.ObjectExpression:
             // delegate object-style tests to the invalid checker
             checkInvalidTest(test, false);
@@ -546,7 +549,7 @@ export default createRule<Options, MessageIds>({
 
               case 'invalid':
                 for (const element of prop.value.elements) {
-                  if (element.type === AST_NODE_TYPES.ObjectExpression) {
+                  if (element?.type === AST_NODE_TYPES.ObjectExpression) {
                     checkInvalidTest(element);
                   }
                 }
@@ -575,7 +578,7 @@ export default createRule<Options, MessageIds>({
             }
 
             for (const errorElement of testProp.value.elements) {
-              if (errorElement.type !== AST_NODE_TYPES.ObjectExpression) {
+              if (errorElement?.type !== AST_NODE_TYPES.ObjectExpression) {
                 continue;
               }
 

--- a/packages/eslint-plugin/src/rules/no-unnecessary-condition.ts
+++ b/packages/eslint-plugin/src/rules/no-unnecessary-condition.ts
@@ -568,9 +568,7 @@ export default createRule<Options, MessageId>({
       return false;
     }
 
-    function isOptionableExpression(
-      node: TSESTree.LeftHandSideExpression,
-    ): boolean {
+    function isOptionableExpression(node: TSESTree.Expression): boolean {
       const type = getNodeType(node);
       const isOwnNullable =
         node.type === AST_NODE_TYPES.MemberExpression

--- a/packages/eslint-plugin/src/rules/prefer-string-starts-ends-with.ts
+++ b/packages/eslint-plugin/src/rules/prefer-string-starts-ends-with.ts
@@ -47,7 +47,7 @@ export default createRule({
      * Check if a given node is a string.
      * @param node The node to check.
      */
-    function isStringType(node: TSESTree.LeftHandSideExpression): boolean {
+    function isStringType(node: TSESTree.Expression): boolean {
       const objectType = typeChecker.getTypeAtLocation(
         service.esTreeNodeToTSNodeMap.get(node),
       );

--- a/packages/eslint-plugin/src/rules/require-array-sort-compare.ts
+++ b/packages/eslint-plugin/src/rules/require-array-sort-compare.ts
@@ -50,7 +50,7 @@ export default util.createRule<Options, MessageIds>({
      * Check if a given node is an array which all elements are string.
      * @param node
      */
-    function isStringArrayNode(node: TSESTree.LeftHandSideExpression): boolean {
+    function isStringArrayNode(node: TSESTree.Expression): boolean {
       const type = checker.getTypeAtLocation(
         service.esTreeNodeToTSNodeMap.get(node),
       );

--- a/packages/typescript-estree/tests/lib/semanticInfo.test.ts
+++ b/packages/typescript-estree/tests/lib/semanticInfo.test.ts
@@ -365,7 +365,7 @@ function testIsolatedFile(
   const declaration = (parseResult.ast.body[0] as TSESTree.VariableDeclaration)
     .declarations[0];
   const arrayMember = (declaration.init! as TSESTree.ArrayExpression)
-    .elements[0];
+    .elements[0]!;
   expect(parseResult).toHaveProperty('services.esTreeNodeToTSNodeMap');
 
   // get corresponding TS node

